packages/oui-dual-list/src/dual-list.controller.js

```diff
@@ -0,0 +1,247 @@
+        this.loadSourceList();
+        this.loadtargetList();
+    }
+
+    getSourceContentHeight () {
+        if (this.bulkActionEnabled) {
+            return { height: "calc(100% - 90px)" };
+        }
+        return { height: "calc(100% - 60px)" };
+    }
+
+    getTargetContentHeight () {
+        if (this.bulkActionEnabled) {
+            return { height: "calc(100% - 30px)" };
+        }
+        return { height: "100%" };
+    }
+
+    onTargetContentClose () {
+        this.targetCloseIconStyle = this.displayNoneStyle;
+        this.targetOpenIconStyle = this.displayBlockStyle;
+        this.targetContentStyle = this.displayFlexStyle;
+    }
+
+    onTargetContentOpen () {
+        this.targetCloseIconStyle = this.displayBlockStyle;
+        this.targetOpenIconStyle = this.displayNoneStyle;
+        this.targetContentStyle = this.displayNoneStyle;
+    }
+
+    onSourceContentClose () {
+        this.sourceCloseIconStyle = this.displayNoneStyle;
+        this.sourceOpenIconStyle = this.displayBlockStyle;
+        this.sourceContentStyle = this.displayFlexStyle;
+    }
+
+    onSourceContentOpen () {
+        this.sourceCloseIconStyle = this.displayBlockStyle;
+        this.sourceOpenIconStyle = this.displayNoneStyle;
+        this.sourceContentStyle = this.displayNoneStyle;
+    }
+
+    getProperty (item) {
+        if (!this.property) {
+            return item;
+        }
+        return this.property.split(".").reduce((prev, curr) => prev ? prev[curr] : undefined, item);
+    }
+
+    isLoading (item) {
+        const uniqueName = this.getProperty(item);
+        if (this.loadingMap[uniqueName]) {
+            return true;
+        }
+        return false;
+    }
+
+    setLoading (item, state) {
+        const uniqueName = this.getProperty(item);
+        this.loadingMap[uniqueName] = state;
+    }
+
+    loadSourceList () {
+        if (this.sourceListLoading) {
+            return this.$q.reject(false);
+        }
+        this.sourceListLoading = true;
+        return this.$q.when(this.sourceList)
+            .then(items => {
+                this.sourceList = items.data ? items.data : items;
+            })
+            .finally(() => {
+                this.sourceListLoading = false;
+            });
+    }
+
+    loadtargetList () {
+        if (this.targetListLoading) {
+            return this.$q.reject(false);
+        }
+        this.targetListLoading = true;
+        return this.$q.when(this.targetList)
+            .then(items => {
+                this.targetList = items.data ? items.data : items;
+            })
+            .finally(() => {
+                this.targetListLoading = false;
+            });
+    }
+
+    add (index, item) {
+        if (this.isLoading(item)) {
+            return;
+        }
+        this.sourceList.splice(index, 1);
+        this.targetList.push(item);
+        if (this.onAdd) {
+            this.setLoading(item, true);
+            this.onAdd({ items: [item] })
```

Comment:
`onAdd`is always called with an object containing a list of items, we could just pass the list ?

Reply:
No. items which are loading should not be moved by clicking on add. That's why I loop through all items and pick items which are not in the loading state.

Reply:
I just mean replacing ```this.onAdd({ items: [item] })``` by ```this.onAdd([item])```

Reply:
the component has to send an array of items as it supports bulk operations too.